### ISSUE TITLE
HVM import tool

### ISSUE THREAD

**Issue by jpouellet** [QubesOS/qubes-issues#2634]
I think it would be greatly beneficial if there were an easy graphical external-disk -> qubes HVM import tool. I'm imagining a button in qubes-manager similar to backup & restore, that would allow you to select a block device (possibly in sys-usb or such) and create an *actually-booting* HVM from it.

The ability to pop in a new (likely larger) drive and import your old installation in a way that it boots and "just works" greatly lowers the barrier for adopting Qubes because it allows the change of user habits and workflows to happen gradually and at the pace the user is comfortable with (as they can still get their work done in the HVM of their old installation if they need to), instead of forcing the switch to Qubes to be something that must be completed over a weekend or not happen at all.

Based on my experience helping linux-friends become qubes-friends, I think this might perhaps be the single most helpful thing we could do for technically competent prospective users.

> "I thought I would be able to just `dd of=root.img`, but several days later I still have no clue why it won't boot" -- a friend who recently tried to install Qubes (where the problem seems likely to have been due to differences in expectations between his UEFI boot sequence & pvgrub, but ran out of time to debug it)

Things such a tool would ideally support:
- source devices attached to non-dom0
- flexible support for custom disk layouts: I've had friends experience problems due to installations on LVM, zvols, various btrfs shenanigans, etc. IMO the ability to expose your volume as a block device in an AppVM (where you can install whatever drivers, etc.) should be the only dependency

Obviously there would be many corner cases, but I think it's worth attempting.

@mfc I'm interested in your thoughts and experience on this as well.
Edit: Oops, maybe I meant @micahflee? Same letters, different order. Regardless... I mean people who have experience helping others start using Qubes.

**Comment by emdete** [comment]
i like the baseline of this issue, this "how do i wrap qubes around my usecases". i doubt the solution mentioned "how do i copy a HVM in" is the desired way to go.

maybe one should create a location where some beginners collect their qubes expiriences? how everyone adapt the system to her needs? from that it could be easier to learn what kind of gui and features are really needed.

**Comment by jpouellet** [comment]
@emdete sure, more formal usability testing would be nice.

Regardless, I've collected enough personal and anecdotal evidence that this is something people want to think it warrants a ticket.

**Comment by andrewdavidwong** [comment]
> maybe one should create a location where some beginners collect their qubes expiriences? how everyone adapt the system to her needs? from that it could be easier to learn what kind of gui and features are really needed.

Might be covered by #1906.

**Comment by emdete** [comment]
> Might be covered by #1906.

that one is still other way around: people say what they think which people potentially [sh|c]ould use qubes.

i am interested in the knowledge of people already _using_ qubes. how do you manage daily tasks. how did you migrate to qubes (i use a double boot setup for now to lean how to adapt my flows)? does video editiing actually work (here: no, qubes makes my box too slow for playback without stutter)? where do you put thousands of images and how do you manage the backup of the newly taken pics (there is currently no usable backup solution in qubes as only full backup of VMs is suppported)? can i use qubes for virtualization purposes to test setups beside security concerns (yes! it works incredible fine!)? can i pimp my setup (yes, everyting in qubes is well separated into GUI components, commandline tools, APIs and such that are usable!)?

all the learning of "coming to qubes" from different people could be collected somewhere to be usable by others. how about a github repo with directories for each user that can be filled with READMEs, config snippets and scripts? (ok, someone has to maintain the PRs...)

**Comment by andrewdavidwong** [comment]
> all the learning of "coming to qubes" from different people could be collected somewhere to be usable by others. how about a github repo with directories for each user that can be filled with READMEs, config snippets and scripts? (ok, someone has to maintain the PRs...)

Many messages of this kind are sent to qubes-users, which automatically "collects" them into a searchable archive. I encourage you (or whoever needs it) to try searching for the information you (or they) are looking for.

**Comment by emdete** [comment]
yes, i looked through that as well. it's a huge collection which is really difficult to consume...

**Comment by andrewdavidwong** [comment]
> yes, i looked through that as well. it's a huge collection which is really difficult to consume...

Ah, so the real problem is not *collecting* such information (in the sense of possessing it and making it available). Rather, the real problem is one of *curation*. I'm afraid that we (the Qubes team) can't help here: We're already understaffed as it is. However, it would be great if *you* could help us with this. :)

**Comment by emdete** [comment]
:) happy to do so:

https://github.com/emdete/Qubes-adventure-blogs

but somehow people must be able to find it. if this kind of thing is wanted it should be linked in a prominent place somewhere.

p.s.: be aware that my README only contains my personal keywords to remind me what i still have to look at... it's just to fill the structure.

**Comment by andrewdavidwong** [comment]
> :) happy to do so:
> 
> https://github.com/emdete/Qubes-adventure-blogs
> 
> but somehow people must be able to find it. if this kind of thing is wanted it should be linked in a prominent place somewhere.
> 
> p.s.: be aware that my README only contains my personal keywords to remind me what i still have to look at... it's just to fill the structure.

Cool, thanks! So, how about this idea: You post this on qubes-users and try to encourage some uptake. If it catches on and people are finding it useful, we can then add a link to the documentation (and of course you can continue to promote it on qubes-users). What do you think?

**Comment by emdete** [comment]
> You post this on qubes-users and try to encourage some uptake. If it catches on and people are finding it useful, we can then add a link to the documentation (and of course you can continue to promote it on qubes-users). What do you think?

i am a really bad missionary (remember my inter-vm-fs where i did not get one reaction ;) ) but can publish it. will do when my text is more comprehensive.

**Comment by grey-olli** [comment]
you may also look over messages w/ 'qubes' tag in my livejournal blog.

**Comment by jpouellet** [comment]
xref https://github.com/QubesOS/qubes-issues/issues/2737#issuecomment-294390220:
> On the other hand, in Qubes 4.0, you can have management VM having power to create VM, adjust properties etc, so all the format parsing can be sandboxed outside of dom0. Besides feature discussed here, it will allow cool things like implementing VM import from almost arbitrary complex format (OVA/OVF?). Without compromising dom0/other VMs.

**Comment by ben-grande** [comment]
I believe this is solved by "Qube Manger -> Advanced -> Boot qube from DISC or block device".

**Comment by andrewdavidwong** [comment]
> I believe this is solved by "Qube Manger -> Advanced -> Boot qube from DISC or block device".

I think they're about two different things:

- Create an empty HVM and boot from an ISO to install an OS (what you posted)
- Take an existing non-empty HVM from outside of your Qubes installation and import it into your Qubes installation (what this issue is about)

**Comment by ben-grande** [comment]
> I think they're about two different things:

Yes, you're right. It seems that this issue is about `qvm-create --root-copy-from|--root-move-from` and there is no issue open about this.
